packages/wrangler/src/index.tsx

```diff
@@ -1119,6 +1119,71 @@ function createCLIParser(argv: string[]) {
         );
       }
 
+      // Let's get all the vars we're going to use into an object here
+      // since we're going to manipulate it later
+      const combinedVars = getVarsForDev(config);
+
+      const bindings = {
+        kv_namespaces: config.kv_namespaces?.map(
+          ({ binding, preview_id, id: _id }) => {
+            // In `dev`, we make folks use a separate kv namespace called
+            // `preview_id` instead of `id` so that they don't
+            // break production data. So here we check that a `preview_id`
+            // has actually been configured.
+            // This whole block of code will be obsoleted in the future
+            // when we have copy-on-write for previews on edge workers.
+            if (!preview_id) {
+              // TODO: This error has to be a _lot_ better, ideally just asking
+              // to create a preview namespace for the user automatically
+              throw new Error(
+                `In development, you should use a separate kv namespace than the one you'd use in production. Please create a new kv namespace with "wrangler kv:namespace create <name> --preview" and add its id as preview_id to the kv_namespace "${binding}" in your wrangler.toml`
+              ); // Ugh, I really don't like this message very much
+            }
+            return {
+              binding,
+              id: preview_id,
+            };
+          }
+        ),
+        // Use a copy of combinedVars since we're modifying it later
+        vars: { ...combinedVars },
+        wasm_modules: config.wasm_modules,
+        text_blobs: config.text_blobs,
+        data_blobs: config.data_blobs,
+        durable_objects: config.durable_objects,
+        r2_buckets: config.r2_buckets?.map(
+          ({ binding, preview_bucket_name, bucket_name: _bucket_name }) => {
+            // same idea as kv namespace preview id,
+            // same copy-on-write TODO
+            if (!preview_bucket_name) {
+              throw new Error(
+                `In development, you should use a separate r2 bucket than the one you'd use in production. Please create a new r2 bucket with "wrangler r2 bucket create <name>" and add its name as preview_bucket_name to the r2_buckets "${binding}" in your wrangler.toml`
+              );
+            }
+            return {
+              binding,
+              bucket_name: preview_bucket_name,
+            };
+          }
+        ),
+        services: config.services,
+        unsafe: config.unsafe?.bindings,
+      };
+
+      // mask anything that was overridden in .dev.vars
+      // so that we don't log potential secrets into the terminal
+      for (const key of Object.keys(config.vars)) {
+        if (combinedVars[key] !== config.vars[key]) {
+          combinedVars[key] = "(hidden)";
+        }
+      }
```

Comment:
I agree the function signature that I suggested is weird. This could be simpler by changing other stuff, like not merging the vars in `getVarsForDev()`. The concern I have is that the work to overwrite these vars with `(hidden)` in this bit of code has nothing to do with setting up the bindings. It is purely an aspect of what to print, which is why I feel uncomfortable doing that work here.

Let's land this as it is then and perhaps I can convince you that we refactor it later.

Reply:
Yeah I can see the concern, and I'm actually sympathetic. Happy to revisit soon. At least this doesn't make things worse than they already are, and fixes the security issue. Could you stamp this so we can land it for now?